src/flexspi/nor_flash.rs

```diff
@@ -0,0 +1,497 @@
+            }
+
+            // Drain the RX queue until the read buffer is full.
+            let read = self.flex_spi.drain_rx_fifo(buffer);
+            if read != buffer.len() {
+                return Err(NotEnoughData {
+                    expected: i * 128 + buffer.len(),
+                    actual: i * 128 + read,
+                }.into())
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Erase a sector of flash memory.
+    ///
+    /// NOTE: The address argument is a physical flash address, not a CPU memory address.
+    ///
+    /// # Safety
+    /// You may not erase flash memory holding code of the current program.
+    ///
+    /// If your program also performs memory mapped access to the erased region,
+    /// you must invalidate the FlexSPI cache and the AHB RX buffer.
+    pub unsafe fn erase_sector(&mut self, address: u32) -> Result<(), WriteError> {
+        // TODO: verify address is aligned to a sector.
+        self.set_and_verify_write_enable()?;
+
+        unsafe {
+            self.flex_spi
+                .configure_command_sequence(CommandSequence {
+                    start: sequence::ERASE_SECTOR,
+                    count: 1,
+                    address,
+                    data_size: 0,
+                    parallel: false,
+                })
+                .unwrap_or_else(|_: InvalidCommandSequence| {
+                    panic!("FlexSPI driver reported invalid command sequence for hard-coded erase sector sequence")
+                });
+                self.flex_spi.trigger_command_and_wait_write()?;
+        }
+        Ok(())
+    }
+
+    /// Erase a block of flash memory.
+    ///
+    /// NOTE: The address argument is a physical flash address, not a CPU memory address.
+    ///
+    /// # Safety
+    /// You may not erase flash memory holding code of the current program.
+    ///
+    /// If your program also performs memory mapped access to the erased region,
+    /// you must invalidate the FlexSPI cache and the AHB RX buffer.
+    pub unsafe fn erase_block(&mut self, address: u32) -> Result<(), WriteError> {
+        // TODO: verify address is aligned to a block.
+        self.set_and_verify_write_enable()?;
+
+        unsafe {
+            self.flex_spi
+                .configure_command_sequence(CommandSequence {
+                    start: sequence::ERASE_BLOCK,
+                    count: 1,
+                    address,
+                    data_size: 0,
+                    parallel: false,
+                })
+                .unwrap_or_else(|_: InvalidCommandSequence| {
+                    panic!("FlexSPI driver reported invalid command sequence for hard-coded erase sector sequence")
+                });
+            self.flex_spi.trigger_command_and_wait_write()?;
+        }
+        Ok(())
+    }
+
+    /// Erase the whole flash chip.
+    ///
+    /// # Safety
+    /// You may not erase flash memory holding code of the current program.
```

Comment:
This will be hard to ensure if we're running off of flash, right? Erasing the entire chip, but can't erase code for the current program. That being said, no idea how you could overcome this one. Perhaps just document that this cannot be callein XIP scenarios?